### DISCUSSION TITLE
Is a data disk size of 20gb required?

### DISCUSSION THREAD

**Discussion by andreialecu** [bottlerocket-os/bottlerocket#4061] [General]
I'm using Bottlerocket with AWS ECS to orchestrate EC2 Spot instances.

I noticed each instance has two EBS volumes, one 2GB and the other 20GB.

My question is if the data volume is not unnecessarily large at 20GB, as this contributes to some costs for our AWS bill.

As a test, I reduced it to 10GB in the block devices configuration of our launch template and everything works just fine.

What am I risking? Can someone provide some guidance?

Thanks!

**Comment by larvacea** [comment]
As you may have guessed already, this is the persistent, read-write storage for your instance. The 2GB volume is read only. We mount /var, /local, /opt, and /mnt on that larger read-write EBS volume as xfs file systems. Of those, I would expect /var/lib/containerd (EKS) or /var/lib/docker (ECS) to be the largest and most variable consumer (but my experience is limited and non-production). The most serious risk would be that the instance would crash or become nonresponsive, as with any disk-completely-full scenario, and the most likely cause would be something in your world causing dockerd to fill its data-root. I'm afraid I don't have any specific, actionable guidance, though.

**Comment by yeazelm** [comment]
>My question is if the data volume is not unnecessarily large at 20GB, as this contributes to some costs for our AWS bill.
> As a test, I reduced it to 10GB in the block devices configuration of our launch template and everything works just fine.

To summarize @larvacea's comments: The root volume must be the 2GB in size, but the secondary volume only _defaults_ to 20GB. If you can run your workload with smaller disks, you can override the default 20GB using block device mappings. In fact, many users of Bottlerocket set the secondary disk to be larger but you can also set it to smaller as you discovered. The primary concern is that this secondary disk holds nearly all variable data: so your logs, container images, etc all default to that volume. The size of this volume really is workload dependent so there isn't a general one-size-fits all recommendation but if you found it works with 10GB, you should be fine as long as you watch how much you are writing to that volume.

**Comment by larvacea** [comment]
The read-only root volume must be at least 2GB in size. I have seen instances where a customer has chosen a larger size, but I'd like to state here that this has no benefit: the instance will not use any extra capacity over the 2GB default. Resizing the data volume up or down does matter. It behaves somewhat like the root volume in a general-purpose Linux distro, and the same considerations apply. It is only somewhat-like, I should say. For instance, in most distros, /etc is part of the root volume and persists between boots, but in Bottlerocket /etc is ephemeral (a tmpfs in-memory filesystem), and is rebuilt during each boot. This surprises me more often than I'd like, even though I've seen how systemd builds it and I should know better.

**Comment by SAAMalusare** [comment]
I have observed another issue, my data volume is running out of space even though I am allocating 30G for data vol.
Error ''The node was low on resource: ephemeral-storage. Threshold quantity: 3210844697, available: 2711696Ki.'"

After investigating we observed that "/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs" takes up 12G, not sure why this is happening as we start fresh instance every morning.

Am I missing any setting?

Thank you in advance :)

**Reply by SAAMalusare** [reply]
ephemeral-storage is 5G

**Comment by sw360cab** [comment]
Hi. going back to this discussion to have a deeper understanding of Bottlerocket OS.
So far, I understood the AMI of the OS comes bundled into two discs:
- 1 disk readonly, 2 GB size, reserved  to Bottlerocket itself
- 1 disk, 20 GB size, where containers images and logs are written to.

I am trying to create a bigger disk to replace or stay aside the previous disk but what I discovered is that
- I don't see a way to shrink / enlarge this disk
- if I add an additional disk it result unused by Bottlerocket OS

What am I missing?

**Reply by yeazelm** [reply]
@sw360cab I'll try to answer your questions below:

> I don't see a way to shrink / enlarge this disk

You would specify a difference size as a [block device mapping](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/block-device-mapping-concepts.html). The root volume is `/dev/xvda` and the data volume is `/dev/xvdb`. `eksctl` has some docs on specifying volumes [here](https://docs.aws.amazon.com/eks/latest/eksctl/nodegroup-additional-volume-mappings.html). Karpenter has [similar docs](https://karpenter.sh/docs/concepts/nodeclasses/#bottlerocket-1) on configuring these volumes.

If you are using Karpenter, something like 

>if I add an additional disk it result unused by Bottlerocket OS

Bottlerocket won't automatically operate additional EBS volumes just by providing them. If you want additional volumes, you should use the EBS CSI Driver to create the right PersistentVolumes and StorageClasses you need.

**Reply by sw360cab** [reply]
> Bottlerocket won't automatically operate additional EBS volumes just by providing them. If you want additional volumes, you should use the EBS CSI Driver to create the right PersistentVolumes and StorageClasses you need.

@yeazelm I see, but actually I was meaning disk for storing basically logs and containers instances.
AFAIK this is in charge to Bottlerocket OS. The parts for storage using EBS CSI, is something I already have in place and works super smoothly from first time I migrated to BottlerocketOS.

**Reply by yeazelm** [reply]
I'm not recommending you add another device, if you want to change the size of the data volume where logs and container images are stored, you should configure the block device mapping for `/dev/xvdb`.

**Reply by sw360cab** [reply]
> You would specify a difference size as a [block device mapping](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/block-device-mapping-concepts.html). The root volume is /dev/xvda and the data volume is /dev/xvdb. eksctl has some docs on specifying volumes [here](https://docs.aws.amazon.com/eks/latest/eksctl/nodegroup-additional-volume-mappings.html). Karpenter has [similar docs](https://karpenter.sh/docs/concepts/nodeclasses/#bottlerocket-1) on configuring these volumes.

I scheduled my nodes using Terraform and I have block-devices setup in the node

```
  block_device_mappings {
    device_name = "/dev/sdf"

    ebs {
      volume_size = 40
      volume_type = "gp3"
      delete_on_termination = true
  }  
}
```

If I use the `admin container` I can even see them as correctly mounted. By the disk is empty and remains as is. and actually Bottlerocket is not using it for node's duties (logs, images and so on) as Iexpected, but it keeps using default data disc (20 GB one)

**Reply by sw360cab** [reply]
> I'm not recommending you add another device, if you want to change the size of the data volume where logs and container images are stored, you should configure the block device mapping for /dev/xvdb.

Sorry, have not read that.... WOW, did not know that, I'll try immediately 💪

**Reply by sw360cab** [reply]
@yeazelm Following your advice.
~~Here is what I got. I think it did not work. What do you think?~~

```
Waiting for admin container to start.......
Entering admin container
          Welcome to Bottlerocket's admin container!
    ╱╲
   ╱┄┄╲   This container provides access to the Bottlerocket host
   │▗▖│   filesystems (see /.bottlerocket/rootfs) and contains common
  ╱│  │╲  tools for inspection and troubleshooting.  It is based on
  │╰╮╭╯│  Amazon Linux 2, and most things are in the same places you
    ╹╹    would find them on an AL2 host.

To permit more intrusive troubleshooting, including actions that mutate the
running state of the Bottlerocket host, we provide a tool called "sheltie"
(`sudo sheltie`).  When run, this tool drops you into a root shell in the
Bottlerocket host's root filesystem.
[root@admin]# sheltie
bash-5.1# lsblk
NAME         MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
nvme1n1      259:0    0   40G  0 disk 
nvme0n1      259:1    0    2G  0 disk 
|-nvme0n1p1  259:3    0    4M  0 part 
|-nvme0n1p2  259:4    0    5M  0 part 
|-nvme0n1p3  259:5    0   40M  0 part /boot
|-nvme0n1p4  259:6    0  920M  0 part 
|-nvme0n1p5  259:7    0   10M  0 part 
|-nvme0n1p6  259:8    0   25M  0 part 
|-nvme0n1p7  259:9    0    5M  0 part 
|-nvme0n1p8  259:10   0   40M  0 part 
|-nvme0n1p9  259:11   0  920M  0 part 
|-nvme0n1p10 259:12   0   10M  0 part 
|-nvme0n1p11 259:13   0   25M  0 part 
|-nvme0n1p12 259:14   0   41M  0 part /var/lib/bottlerocket
`-nvme0n1p13 259:15   0    1M  0 part 
nvme2n1      259:2    0   20G  0 disk 
`-nvme2n1p1  259:17   0   20G  0 part /var
                                      /opt
                                      /mnt
                                      /local
```

**Reply by sw360cab** [reply]
@yeazelm my bad, it it actually working!

**Reply by yeazelm** [reply]
Glad you got it working @sw360cab!
